bootstrap/src/basis/file.c

```diff
@@ -53,82 +60,242 @@ hm_basis_file_finalize_result(int result) {
 }
 
 CAMLprim value
-hm_basis_file_of_path_inner(value a_flag, value a_mode, value a_bytes) {
-    size_t flag = Long_val(a_flag);
-    size_t mode = Int64_val(a_mode);
+hm_basis_file_stdin_inner(value a_unit) {
+    return caml_copy_int64(STDIN_FILENO);
+}
+
+CAMLprim value
+hm_basis_file_stdout_inner(value a_unit) {
+    return caml_copy_int64(STDOUT_FILENO);
+}
+
+CAMLprim value
+hm_basis_file_stderr_inner(value a_unit) {
+    return caml_copy_int64(STDERR_FILENO);
+}
+
+CAMLprim value
+hm_basis_file_write_inner(value a_bytes, value a_fd) {
     uint8_t *bytes = (uint8_t *)Bytes_val(a_bytes);
     size_t n = caml_string_length(a_bytes);
+    int fd = Int64_val(a_fd);
 
-    int flags = flags_of_hemlock_file_flag[flag];
-    int result = open(bytes, flags, mode);
+    assert(n > 0);
+    int result = write(fd, bytes, n);
 
     return hm_basis_file_finalize_result(result);
 }
 
 CAMLprim value
-hm_basis_file_stdin_inner(value a_unit) {
-    return caml_copy_int64(STDIN_FILENO);
+hm_basis_file_seek_inner(value a_i, value a_fd) {
+    size_t i = Int64_val(a_i);
+    int fd = Int64_val(a_fd);
+
+    return hm_basis_file_finalize_result(lseek(fd, i, SEEK_CUR));
 }
 
 CAMLprim value
-hm_basis_file_stdout_inner(value a_unit) {
-    return caml_copy_int64(STDOUT_FILENO);
+hm_basis_file_seek_hd_inner(value a_i, value a_fd) {
+    size_t i = Int64_val(a_i);
+    int fd = Int64_val(a_fd);
+
+    return hm_basis_file_finalize_result(lseek(fd, i, SEEK_SET));
 }
 
 CAMLprim value
-hm_basis_file_stderr_inner(value a_unit) {
-    return caml_copy_int64(STDERR_FILENO);
+hm_basis_file_seek_tl_inner(value a_i, value a_fd) {
+    size_t i = Int64_val(a_i);
+    int fd = Int64_val(a_fd);
+
+    return hm_basis_file_finalize_result(lseek(fd, i, SEEK_END));
 }
 
+// hm_basis_file_user_data_decref: Basis.File.{Open|Close|Read|Write}.t >{os}-> unit
+value
+hm_basis_file_user_data_decref(value a_user_data) {
+    hm_user_data_decref((hm_user_data_t *)Int64_val(a_user_data));
+
+    return Val_unit;
+}
+
+// hm_basis_file_user_data_pp: Basis.File.t -> Basis.File.{Open|Close|Read|Write}.t >{os}-> unit
 CAMLprim value
-hm_basis_file_close_inner(value a_fd) {
+hm_basis_file_user_data_pp(value a_fd, value a_user_data) {
+    hm_user_data_t *user_data = (hm_user_data_t *)Int64_val(a_user_data);
     int fd = Int64_val(a_fd);
 
-    return hm_basis_file_finalize_result(close(fd));
+    hm_user_data_pp(fd, user_data);
+
+    return Val_unit;
+}
+
+// hm_basis_file_complete_inner: Basis.File.{Open|Close|Write}.t >{os}-> int
```

Comment:
It occurred to me that this signature can't be correct. If `complete` simply had an `>{os}->` effect, it'd be valid to package it up in a lazy closure and stuff it onto the global heap, allowing any other actor to perform the completion. That's bad, because a submission sets up space within the associated actor for the completion result. I was a little worried that this implied we might need something like a `>{complete}->` effect to prevent this, but I think the answer might be to just make `Basis.File.{Open|Close|Write}.t` mutable. It'd make it valid for the mutable types to exist in an immutable lazy closure only if it was created and completed within a single `Lazy.force`. All that would be done within the same actor.

This isn't anything new. We discussed it in the past. I just realized I'd gotten these signature comments wrong.